### ISSUE TITLE
SMILE to memory tensor

### ISSUE THREAD

**Issue by gambalab** [MolecularAI/Chemformer#29]
I've been working on using the encoder of a pre-trained transformer model to extract the 2D  memory tensor representation of a SMILES molecule. After some initial challenges, I've successfully configured the model to generate these memory tensors.

It appears that the encoder expects an input vector where each element represents the IDs of the corresponding token in the SMILES sequence. To accommodate variations in SMILES length, the input vector is padded to a fixed length of 512 using the ID of the padding character (i.e., 0).

However, I'm still unsure about the specific role of the padding mask vector. Does it indicate true values at padded positions or where the actual SMILES sequence resides?

**Comment by EBjerrum** [comment]
It prevents attention to padding tokens after the SMILES sequence. The attention is removed from these and thus they do not contribute to the pooling layer information. Its similar to the ones used in the pytorch transformer module: https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html#torch.nn.Transformer

**Comment by gambalab** [comment]
Thank you for explaining the role of padding masks. To ensure I'm using it correctly in Chemformer, could you clarify if the mask ignores positions with True or False values? I understand this behavior can vary, and finding the specific approach in Chemformer would be highly valuable.

From what you linked is reported "the positions of the mask with the value of True will be ignored while the position with the value of False will be unchanged". Is this the case also for Chemformer?

**Comment by EBjerrum** [comment]
I'm not sure why you don't use the included tokenizer class from chemformer https://github.com/MolecularAI/Chemformer/blob/main/molbart/modules/tokenizer.py or look at the included dataloaders and datamodules, it should already delivering all you need for the model? Sounds like you are reimplementing a lot of stuff that is already there, and risk doing things differently than for the pretrained models?

**Comment by gambalab** [comment]
that is how I implemented it:
```python 
import csv
import torch
import pandas as pd
import numpy as np
from molbart.modules.decoder import DecodeSampler
from torch.nn.functional import pad
import molbart.modules.util as util
from molbart.models.transformer_models import BARTModel
from molbart.modules.tokenizer import ChemformerTokenizer

vocabulary_path = util.DEFAULT_VOCAB_PATH
model_path = './models/combined-large/step=1000000.ckpt'


tokenizer = ChemformerTokenizer(filename=util.DEFAULT_VOCAB_PATH)
vocabulary_size = len(tokenizer)

sampler = DecodeSampler(tokenizer, util.DEFAULT_MAX_SEQ_LEN)

model = BARTModel.load_from_checkpoint(model_path, decode_sampler=sampler, vocabulary_size=vocabulary_size)
model.eval()

input_csv_path = "~/mydata/drug_with_SMILES_table.csv"
output_csv_path = "~/SMILES_embedding.csv"
df = pd.read_csv(input_csv_path)

for i in range(df.shape[0]):
    print(i)
    example_smile = [df.loc[i, "canonical_smiles"]]
    n_sample = len(example_smile)
    seq_len = util.DEFAULT_MAX_SEQ_LEN
    pad_token = tokenizer.vocabulary[tokenizer.special_tokens['pad']]
    tokenized_input = tokenizer.encode(example_smile)
    inputs = []
    masks = []
    for s in tokenized_input:
        mask = torch.zeros(s.shape, dtype=torch.bool)
        inputs.append(pad(s, pad=(0, seq_len - s.shape[0]), value=pad_token))
        masks.append(pad(mask, pad=(0, seq_len - s.shape[0]), value=True))


    batch = {
        "encoder_input": torch.reshape(torch.cat(inputs), (n_sample, seq_len)).T,
        "encoder_pad_mask": torch.reshape(torch.cat(masks), (n_sample, seq_len)).T,
    }

    outputs = model.encode(batch)

    mask = ~batch['encoder_pad_mask'][:, 0]
    embedded = outputs[mask, 0, :]
    embedded = torch.mean(embedded, dim=0)
    embedded = embedded.detach().numpy()
    embedded = np.transpose(embedded)
    with open(output_csv_path, 'a') as f:
        writer = csv.writer(f, delimiter=',', lineterminator='\n')
        writer.writerow(embedded)
```

**Comment by EBjerrum** [comment]
Yes, you are trying to handle the vectorization manually instead of using the infrastructure from chemformer in the form of datamodules and dataloaders. Take a look in this directory: 

https://github.com/MolecularAI/Chemformer/tree/main/molbart/modules/data

 Theres a base datamodule, and subclasses for specific datasets. Yours will be similar to the ones in mol_data I think (Its been a while). Then when the datamodule is ready, you can directly use it for pytorch lightning training, or in your instance, simply get the dataloader for the full dataset, and iterate through the batches, which will come in a form ready to feed into the model.
